internal/controllers/machine/machine_controller_inplace_update.go

```diff
@@ -0,0 +1,240 @@
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/core/v1beta2"
+	runtimehooksv1 "sigs.k8s.io/cluster-api/api/runtime/hooks/v1alpha1"
+	"sigs.k8s.io/cluster-api/feature"
+	"sigs.k8s.io/cluster-api/internal/hooks"
+)
+
+// reconcileInPlaceUpdate handles the in-place update workflow for a Machine.
+func (r *Reconciler) reconcileInPlaceUpdate(ctx context.Context, s *scope) (ctrl.Result, error) {
+	if !feature.Gates.Enabled(feature.InPlaceUpdates) {
+		return ctrl.Result{}, nil
+	}
+
+	log := ctrl.LoggerFrom(ctx)
+
+	machineAnnotations := s.machine.GetAnnotations()
+	_, inPlaceUpdateInProgress := machineAnnotations[clusterv1.InPlaceUpdateInProgressAnnotation]
+	hasUpdateMachinePending := hooks.IsPending(runtimehooksv1.UpdateMachine, s.machine)
+
+	if !inPlaceUpdateInProgress {
+		// Clean up any orphaned pending hooks and annotations before exiting.
+		// This can happen if the in-place update annotation was removed from Machine
+		// but the UpdateMachine hook is still pending or annotations are still on InfraMachine/BootstrapConfig.
+		if hasUpdateMachinePending {
+			log.Info("In-place update annotation removed but UpdateMachine hook still pending, cleaning up orphaned hook and annotations")
+			if err := r.completeInPlaceUpdate(ctx, s); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to clean up orphaned UpdateMachine hook and annotations")
+			}
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	// If hook is not pending, we're waiting for the owner controller to mark it as pending.
+	if !hasUpdateMachinePending {
+		log.Info("In-place update annotations are set, waiting for UpdateMachine hook to be marked as pending")
+		return ctrl.Result{}, nil
+	}
+
+	if s.infraMachine == nil {
+		s.updatingReason = clusterv1.MachineUpdateFailedReason
+		s.updatingMessage = "In-place update not possible: InfraMachine not found"
+		return ctrl.Result{}, errors.New("in-place update failed: InfraMachine not found")
+	}
+
+	infraReady := r.isInfraMachineReadyForUpdate(s)
+	bootstrapReady := r.isBootstrapConfigReadyForUpdate(s)
+
+	if !infraReady || !bootstrapReady {
+		log.Info("Waiting for InfraMachine and BootstrapConfig to be marked for in-place update")
+		return ctrl.Result{}, nil
+	}
+
+	if hasUpdateMachinePending {
+		log.Info("UpdateMachine hook is pending, calling runtime hook")
+		result, err := r.callUpdateMachineHook(ctx, s)
+		if err != nil {
+			s.updatingReason = clusterv1.MachineUpdateFailedReason
+			s.updatingMessage = fmt.Sprintf("UpdateMachine hook failed: %v", err)
+			return ctrl.Result{}, err
+		}
+
+		if result.RequeueAfter > 0 {
+			s.updatingReason = clusterv1.MachineWaitingForUpdateMachineHookReason
+			s.updatingMessage = "UpdateMachine hook in progress"
```

Comment:
Would be nice if we can include the response.Message in this message
(one option would be to set this inside callUpdateMachineHook)